### PR TITLE
Dice loss

### PR DESCRIPTION
**Pull request by myychal**
First pull with the implementation of DICE loss for one class.

### CODE REVIEW

File: brats/losses.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            prediction: output of a network, expected to be already binarized.
+                Dimensions - (Batch, Channels, Height, Width)
+            target: labels.
+                Dimensions - (Batch, Channels, Height, Width)
+
+        Returns:
+            torch.Tensor: Computed Dice Loss
+        """
+        all_but_batch_dims = list(range(1, target.dim()))
+        prediction = prediction.contiguous()
```
Comment (khrynczenko) [inline brats/losses.py]: Are we sure the tensors that get into the loss function aren't contiguous? Even when they aren't do we want to make them so? I think that this requires allocating new memory and copying all the values under consideration and that might be actually more costly that working on non-contiguous tensors.

File: brats/losses.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    Compute Dice Loss for single class.
+    The loss will be averaged across channels.
+    """
+
+    def __init__(self, epsilon: float = 1e-5):
+        """
+        Args:
+            epsilon: smooth factor to prevent division by zero
+        """
+        super(DiceLossOneClass, self).__init__()
```
Comment (khrynczenko) [inline brats/losses.py]: super().__init__() is enough if you don't consider python 2.x

File: tests/test_losses.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import torch
+import pytest
+import numpy as np
+
+from brats.losses import DiceLossOneClass
+
+BATCH_DIMS = (2, 5, 5, 5)
+
+
+class TestDiceLossOneClass:
```
Comment (khrynczenko) [inline tests/test_losses.py]: I think one more test with real values should be performed. Calculate something for small tensors and see if its correct. Cases for ideal and worst fit only might give you a false confidence of properly working function.

File: brats/losses.py (hunk 1 above)
Comment (khrynczenko) [reply brats/losses.py]: Yeah, you could bring this back when you will be training the network and see if there is a execution time difference between case with and without. Maybe it would provide enough info to decide on the matter.

File: tests/test_losses.py (hunk 3 above)
Comment (khrynczenko) [reply tests/test_losses.py]: Great!

File: brats/losses.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        """
+        Args:
+            prediction: output of a network, expected to be already binarized.
+                Dimensions - (Batch, Channels, Height, Width)
+            target: labels.
+                Dimensions - (Batch, Channels, Height, Width)
+
+        Returns:
+            torch.Tensor: Computed Dice Loss
+        """
```
[the remark this answers is not in the document]
Comment (szympie) [reply brats/losses.py]: I think it should comply to the NCDWH, otherwise we would need to do reshaping every time we calculate a loss. I suggest making It work with NCDWH with one channel and raising an exception if the input is larger

[the remark this answers is not in the document]
Comment (myychal) [reply brats/losses.py]: Added Channel dimension

File: brats/losses.py (hunk 1 above)
Comment (szympie) [reply brats/losses.py]: I've added new issue for this, otherwise we would forget about it #23

File: brats/losses.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        all_but_batch_dims = list(range(1, target.dim()))
+        prediction = prediction.contiguous()
+        target = target.contiguous()
+        intersection = (prediction * target).sum(dim=all_but_batch_dims)
```
[the remark this answers is not in the document]
Comment (szympie) [reply brats/losses.py]: I think it would be better if it was called `test_if_raises_on_nonbinarized_input`. I'm sorry for zamieszanie 😅 You can merge it after this change 🙂
